File: mvpLibrary/src/main/java/cn/cbsd/mvplibrary/kit/DateTimePickerUtil.java
package cn.cbsd.mvplibrary.kit;

import android.app.Activity;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;

import com.bigkoo.pickerview.builder.TimePickerBuilder;
import com.bigkoo.pickerview.listener.OnTimeSelectListener;
import com.bigkoo.pickerview.view.TimePickerView;

import java.util.Calendar;
import java.util.Date;


/**
 * 时间选择器
 *
 * @author zhenyanjun
 * @date 2018/10/23 09:48
 */

public class DateTimePickerUtil {

    /**
     * 同时选择日期与时间 记住选择的日期
     */
    public static void dateTimePicker(Activity activity, final TextView target, final String pattern, Calendar selectDate,
                                      Calendar startDate, Calendar endDate) {
        if (startDate == null) {
            startDate = Calendar.getInstance();
            startDate.set(Calendar.YEAR, 1900);
        }
        if (endDate == null) {
            endDate = Calendar.getInstance();
            endDate.set(Calendar.YEAR, 2100);
        }

        TimePickerView timePickerView = new TimePickerBuilder(activity, new OnTimeSelectListener() {
            @Override
            public void onTimeSelect(Date date, View v) {
                String dateStr;
                if (TextUtils.isEmpty(pattern)) {
                    dateStr = DateUtil.getFormatDateTime(date);
                } else {
                    dateStr = DateUtil.getFormatDate(date, pattern);
                }
                target.setText(dateStr);
            }
        })
                .setType(new boolean[]{true, true, true, true, true, true})
                .setRangDate(startDate, endDate)
                .isCyclic(true)
                .build();
        //默认
        timePickerView.setDate(selectDate);
        timePickerView.show(target);
    }

    public static void dateTimePicker(Activity activity, TextView target) {
        dateTimePicker(activity, target, "", Calendar.getInstance(), null, null);
    }

    public static void dateTimePicker(Activity activity, TextView target, String pattern) {
        dateTimePicker(activity, target, pattern, Calendar.getInstance(), null, null);
    }

    public static void dateTimePicker(Activity activity, final TextView target, Calendar startDate, Calendar endDate) {
        dateTimePicker(activity, target, "", Calendar.getInstance(), startDate, endDate);
    }

    public static void dateTimePicker(Activity activity, final TextView target, Calendar selectDate, Calendar startDate, Calendar endDate) {
        dateTimePicker(activity, target, "", selectDate, startDate, endDate);
    }

    /**
     * 兼容旧版本
     */
    public static void dateTimePicker(Activity activity, final TextView target, long minDate, long maxDate) {
        Calendar now = Calendar.getInstance();
        Calendar startDate = Calendar.getInstance();
        if (minDate != 0) {
            startDate.setTimeInMillis(minDate);
        } else {
            //从1900开始
            startDate.set(Calendar.YEAR, 1900);
        }
        Calendar endDate = Calendar.getInstance();
        if (maxDate != 0) {
            endDate.setTimeInMillis(maxDate);
        } else {
            //从2100结束
            endDate.set(Calendar.YEAR, 2100);
        }

        TimePickerView timePickerView = new TimePickerBuilder(activity, new OnTimeSelectListener() {
            @Override
            public void onTimeSelect(Date date, View v) {
                target.setText(DateUtil.getFormatDateTime(date));
            }
        })
                .setType(new boolean[]{true, true, true, true, true, true})
                .setRangDate(startDate, endDate)
                .isCyclic(true)
                .build();
        //默认
        timePickerView.setDate(now);
        timePickerView.show(target);
    }

    /**
     * 选择日期
     */
    public static void datePicker(Activity activity, final TextView target, final String pattern, Calendar selectDate,
                                  Calendar startDate, Calendar endDate, boolean isCycle) {
        if (startDate == null) {
            startDate = Calendar.getInstance();
            startDate.set(Calendar.YEAR, 1900);
        }
        if (endDate == null) {
            endDate = Calendar.getInstance();
            endDate.set(Calendar.YEAR, 2100);
        }
        TimePickerView timePickerView = new TimePickerBuilder(activity, new OnTimeSelectListener() {
            @Override
            public void onTimeSelect(Date date, View v) {
                String dateStr;
                if (TextUtils.isEmpty(pattern)) {
                    dateStr = DateUtil.getFormatDate(date);
                } else {
                    dateStr = DateUtil.getFormatDate(date, pattern);
                }
                target.setText(dateStr);
            }
        })
                .setType(new boolean[]{true, true, true, false, false, false})
                .setRangDate(startDate, endDate)
                .isCyclic(isCycle)
                .build();
        //默认
        timePickerView.setDate(selectDate);
        timePickerView.show(target);
    }

    public static void datePicker(Activity activity, TextView target) {
        datePicker(activity, target, "", Calendar.getInstance(), null, null,true);
    }

    public static void datePicker(Activity activity, TextView target, String pattern) {
        datePicker(activity, target, pattern, Calendar.getInstance(), null, null,true);
    }

    public static void datePicker(Activity activity, TextView target, Calendar startDate, Calendar endDate) {
        datePicker(activity, target, "", Calendar.getInstance(), startDate, endDate,true);
    }

    public static void datePicker(Activity activity, TextView target, Calendar startDate, Calendar endDate, boolean isCycle) {
        datePicker(activity, target, "", Calendar.getInstance(), startDate, endDate ,isCycle);
    }

    /**
     * 兼容旧版本
     */
    public static void datePicker(Activity activity, final TextView target, long minDate, long maxDate) {
        Calendar startDate = Calendar.getInstance();
        Calendar endDate = Calendar.getInstance();
        if (minDate != 0) {
            startDate.setTimeInMillis(minDate);
        } else {
            //从1900开始
            startDate.set(Calendar.YEAR, 1900);
        }
        if (maxDate != 0) {
            endDate.setTimeInMillis(maxDate);
        } else {
            //从2100结束
            endDate.set(Calendar.YEAR, 2100);
        }

        TimePickerView timePickerView = new TimePickerBuilder(activity, new OnTimeSelectListener() {
            @Override
            public void onTimeSelect(Date date, View v) {
                target.setText(DateUtil.getFormatDate(date));
            }
        })
                .setType(new boolean[]{true, true, true, false, false, false})//年月日
                .setRangDate(startDate, endDate)
                .isCyclic(true)
                .build();
        //默认
        timePickerView.setDate(Calendar.getInstance());
        timePickerView.show(target);
    }


    /**
     * 选择时间
     */
    public static void timePicker(Activity activity, final TextView target) {
        TimePickerView timePickerView = new TimePickerBuilder(activity, new OnTimeSelectListener() {
            @Override
            public void onTimeSelect(Date date, View v) {
                target.setText(DateUtil.getFormatTime(date));
            }
        })
                .setType(new boolean[]{false, false, false, true, true, true})//时分秒
                .isCyclic(true)
                .build();
        timePickerView.show(target);
    }

}

File: mvpLibrary/src/main/java/cn/cbsd/mvplibrary/kit/DateTimePickerUtil.kt
package cn.cbsd.mvplibrary.kit

import android.app.Activity
import android.text.TextUtils
import android.widget.TextView
import com.bigkoo.pickerview.builder.TimePickerBuilder
import com.bigkoo.pickerview.listener.OnTimeSelectListener
import java.util.*


/**
 * 时间选择器
 *
 * @author zhenyanjun
 * @date 2018/10/23 09:48
 */

object DateTimePickerUtil {

    /**
     * 同时选择日期与时间 记住选择的日期
     */
    @JvmStatic
    @JvmOverloads
    fun dateTimePicker(activity: Activity, target: TextView, pattern: String = "", selectDate: Calendar = Calendar.getInstance(),
                       startDate: Calendar? = null, endDate: Calendar? = null) {
        var startDate = startDate
        var endDate = endDate
        if (startDate == null) {
            startDate = Calendar.getInstance()
            startDate!!.set(Calendar.YEAR, 1900)
        }
        if (endDate == null) {
            endDate = Calendar.getInstance()
            endDate!!.set(Calendar.YEAR, 2100)
        }

        val timePickerView = TimePickerBuilder(activity, OnTimeSelectListener { date, v ->
            val dateStr: String
            if (TextUtils.isEmpty(pattern)) {
                dateStr = DateUtil.getFormatDateTime(date)
            } else {
                dateStr = DateUtil.getFormatDate(date, pattern)
            }
            target.text = dateStr
        })
                .setType(booleanArrayOf(true, true, true, true, true, true))
                .setRangDate(startDate, endDate)
                .isCyclic(true)
                .build()
        //默认
        timePickerView.setDate(selectDate)
        timePickerView.show(target)
    }

    @JvmStatic
    fun dateTimePicker(activity: Activity, target: TextView, startDate: Calendar, endDate: Calendar) {
        dateTimePicker(activity, target, "", Calendar.getInstance(), startDate, endDate)
    }

    @JvmStatic
    fun dateTimePicker(activity: Activity, target: TextView, selectDate: Calendar, startDate: Calendar, endDate: Calendar) {
        dateTimePicker(activity, target, "", selectDate, startDate, endDate)
    }

    /**
     * 兼容旧版本
     */
    @JvmStatic
    fun dateTimePicker(activity: Activity, target: TextView, minDate: Long, maxDate: Long) {
        val now = Calendar.getInstance()
        val startDate = Calendar.getInstance()
        if (minDate != 0L) {
            startDate.timeInMillis = minDate
        } else {
            //从1900开始
            startDate.set(Calendar.YEAR, 1900)
        }
        val endDate = Calendar.getInstance()
        if (maxDate != 0L) {
            endDate.timeInMillis = maxDate
        } else {
            //从2100结束
            endDate.set(Calendar.YEAR, 2100)
        }

        val timePickerView = TimePickerBuilder(activity, OnTimeSelectListener { date, v -> target.text = DateUtil.getFormatDateTime(date) })
                .setType(booleanArrayOf(true, true, true, true, true, true))
                .setRangDate(startDate, endDate)
                .isCyclic(true)
                .build()
        //默认
        timePickerView.setDate(now)
        timePickerView.show(target)
    }

    /**
     * 选择日期
     */
    @JvmStatic
    @JvmOverloads
    fun datePicker(activity: Activity, target: TextView, pattern: String = "", selectDate: Calendar = Calendar.getInstance(),
                   startDate: Calendar? = null, endDate: Calendar? = null, isCycle: Boolean = true) {
        var startDate = startDate
        var endDate = endDate
        if (startDate == null) {
            startDate = Calendar.getInstance()
            startDate!!.set(Calendar.YEAR, 1900)
        }
        if (endDate == null) {
            endDate = Calendar.getInstance()
            endDate!!.set(Calendar.YEAR, 2100)
        }
        val timePickerView = TimePickerBuilder(activity, OnTimeSelectListener { date, v ->
            val dateStr: String
            if (TextUtils.isEmpty(pattern)) {
                dateStr = DateUtil.getFormatDate(date)
            } else {
                dateStr = DateUtil.getFormatDate(date, pattern)
            }
            target.text = dateStr
        })
                .setType(booleanArrayOf(true, true, true, false, false, false))
                .setRangDate(startDate, endDate)
                .isCyclic(isCycle)
                .build()
        //默认
        timePickerView.setDate(selectDate)
        timePickerView.show(target)
    }

    @JvmStatic
    fun datePicker(activity: Activity, target: TextView, startDate: Calendar, endDate: Calendar) {
        datePicker(activity, target, "", Calendar.getInstance(), startDate, endDate, true)
    }

    @JvmStatic
    fun datePicker(activity: Activity, target: TextView, startDate: Calendar, endDate: Calendar, isCycle: Boolean) {
        datePicker(activity, target, "", Calendar.getInstance(), startDate, endDate, isCycle)
    }

    /**
     * 兼容旧版本
     */
    @JvmStatic
    fun datePicker(activity: Activity, target: TextView, minDate: Long, maxDate: Long) {
        val startDate = Calendar.getInstance()
        val endDate = Calendar.getInstance()
        if (minDate != 0L) {
            startDate.timeInMillis = minDate
        } else {
            //从1900开始
            startDate.set(Calendar.YEAR, 1900)
        }
        if (maxDate != 0L) {
            endDate.timeInMillis = maxDate
        } else {
            //从2100结束
            endDate.set(Calendar.YEAR, 2100)
        }

        val timePickerView = TimePickerBuilder(activity, OnTimeSelectListener { date, v -> target.text = DateUtil.getFormatDate(date) })
                .setType(booleanArrayOf(true, true, true, false, false, false))//年月日
                .setRangDate(startDate, endDate)
                .isCyclic(true)
                .build()
        //默认
        timePickerView.setDate(Calendar.getInstance())
        timePickerView.show(target)
    }


    /**
     * 选择时间
     */
    @JvmStatic
    fun timePicker(activity: Activity, target: TextView) {
        val timePickerView = TimePickerBuilder(activity, OnTimeSelectListener { date, v -> target.text = DateUtil.getFormatTime(date) })
                .setType(booleanArrayOf(false, false, false, true, true, true))//时分秒
                .isCyclic(true)
                .build()
        timePickerView.show(target)
    }

    /**
     * 选择月份
     */
    @JvmStatic
    fun monthPicker(activity: Activity, target: TextView, pattern: String = "", selectDate: Calendar = Calendar.getInstance(),
                   startDate: Calendar? = null, endDate: Calendar? = null, isCycle: Boolean = true) {
        var startDate = startDate
        var endDate = endDate
        if (startDate == null) {
            startDate = Calendar.getInstance()
            startDate!!.set(Calendar.YEAR, 1900)
        }
        if (endDate == null) {
            endDate = Calendar.getInstance()
            endDate!!.set(Calendar.YEAR, 2100)
        }
        val timePickerView = TimePickerBuilder(activity, OnTimeSelectListener { date, v ->
            val dateStr: String
            if (TextUtils.isEmpty(pattern)) {
                dateStr = DateUtil.getFormatDate(date)
            } else {
                dateStr = DateUtil.getFormatDate(date, pattern)
            }
            target.text = dateStr
        })
                .setType(booleanArrayOf(true, true, false, false, false, false))
                .setRangDate(startDate, endDate)
                .isCyclic(isCycle)
                .build()
        //默认
        timePickerView.setDate(selectDate)
        timePickerView.show(target)
    }

}